datafusion/core/src/lib.rs

```diff
@@ -14,6 +14,7 @@
 // KIND, either express or implied.  See the License for the
 // specific language governing permissions and limitations
 // under the License.
+#![feature(generic_associated_types)]
```

Comment:
Unless I'm mistaken this will prevent compilation on Rust stable, which is likely a non-started I'm afraid

Reply:
I am not sure about the version requirement of datafusion, but if we need a pluggable async executor, it should be better to have a higher rank type support, or we have to take some runtime overheads like Pin<Box<Output>>